src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -395,90 +410,97 @@ PayChanClaim::doApply()
     Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
     auto const slep = ctx_.view ().peek (k);
     if (!slep)
-        return tecNO_TARGET;
+        return tecNO_ENTRY;
 
-    AccountID const src = (*slep)[sfAccount];
-    AccountID const dst = (*slep)[sfDestination];
-    AccountID const txAccount = ctx_.tx[sfAccount];
+    auto chanMembers = slep->getFieldArray (sfChannelMembers);
+    auto const& txClaims = ctx_.tx.getFieldArray (sfChannelClaims);
+    bool const thirdPartySubmission =
+        chanMembers[0].getAccountID (sfAccount) != ctx_.tx[sfAccount] &&
+        chanMembers[1].getAccountID (sfAccount) != ctx_.tx[sfAccount];
 
     auto const curExpiration = (*slep)[~sfExpiration];
+    if (curExpiration)
     {
-        auto const cancelAfter = (*slep)[~sfCancelAfter];
         auto const closeTime =
             ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
-        if ((cancelAfter && closeTime >= *cancelAfter) ||
-            (curExpiration && closeTime >= *curExpiration))
+        if (closeTime >= *curExpiration)
             return closeChannel (
                 slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
-    }
 
-    if (txAccount != src && txAccount != dst)
+        // PayChanClaim without a claim does nothing if
+        // the channel has a pending expiration
+        if (txClaims.empty ())
+            return tecNO_PERMISSION;
+    }
+    // third party must have claim if the channel has not expired
+    else if (thirdPartySubmission && txClaims.empty ())
         return tecNO_PERMISSION;
 
-    if (ctx_.tx[~sfBalance])
+    bool submittedOwnClaim = false;
+
+    if (! txClaims.empty ())
     {
-        auto const chanBalance = slep->getFieldAmount (sfBalance).xrp ();
-        auto const chanFunds = slep->getFieldAmount (sfAmount).xrp ();
-        auto const reqBalance = ctx_.tx[sfBalance].xrp ();
+        std::vector<STAmount> amounts {
+            chanMembers[0][sfAmount],
+            chanMembers[1][sfAmount]};
 
-        if (txAccount == dst && !ctx_.tx[~sfSignature])
-            return temBAD_SIGNATURE;
+        std::vector<std::uint32_t> sequences {
+            chanMembers[0][sfSequence],
+            chanMembers[1][sfSequence]};
 
-        if (ctx_.tx[~sfSignature])
+        for (auto const& txClaim : txClaims)
```

Comment:
Calling this the "sequence" appears to be causing confusion. The sequence is not incremented in every claim. It is only incremented when netting/resetting the two payment/claim amounts. If one side's claim amount is up to 1 million XRP and the other is at 1.1 million XRP, the latter can send a claim for 100,000 XRP with a higher sequence. Resetting the amounts is useful for (1) amounts that grow unreasonably large and (2) channel volume privacy.